charger/openevse.go

```diff
@@ -0,0 +1,430 @@
+			return "", false, fmt.Errorf("invalid response from RAPI command %s: %s", command, string(bodyBytes))
+		}
+
+		if matches[1] == "OK" {
+			return strings.TrimSpace(matches[2]), true, nil
+		} else {
+			return "", false, nil
+		}
+	}
+
+	return "", false, fmt.Errorf("invalid response from RAPI command %s: %s, code %d", command, resp.Header.Get("Content-Type"), resp.StatusCode)
+}
+
+func (c *OpenEVSE) SetManualOverride(enable bool) error {
+	state := openevse.SetManualOverrideJSONBodyState("disabled")
+	if enable {
+		state = "active"
+	}
+
+	data := openevse.SetManualOverrideJSONRequestBody{
+		State: &state,
+	}
+
+	ctx, cancel := c.requestContextWithTimeout()
+	defer cancel()
+	_, err := c.api.SetManualOverrideWithResponse(ctx, data)
+
+	return err
+}
+
+func (c *OpenEVSE) Status() (api.ChargeStatus, error) {
+	ctx, cancel := c.requestContextWithTimeout()
+	defer cancel()
+	resp, err := c.api.GetStatusWithResponse(ctx)
+	if err != nil {
+		return api.StatusNone, err
+	}
+
+	/*
+		0: "unknown",
+		1: "not connected",
+		2: "connected",
+		3: "charging",
+		4: "vent required",
+		5: "diode check failed",
+		6: "gfci fault",
+		7: "no ground",
+		8: "stuck relay",
+		9: "gfci self-test failure",
+		10: "over temperature",
+		11: "over current",
+		254: "sleeping",
+		255: "disabled"
+	*/
+
+	state := *resp.JSON200.State
+	vehicleConnected := *resp.JSON200.Vehicle != 0
+
+	switch state {
+	case 1:
+		return api.StatusA, nil
+	case 2, 254, 255:
+		if vehicleConnected {
+			return api.StatusB, nil
+		} else {
+			return api.StatusA, nil
+		}
+	case 3:
+		return api.StatusC, nil
+	case 4:
+		return api.StatusD, nil
+	case 5, 6, 7, 8, 9, 10, 11:
+		return api.StatusF, nil
+	default:
+		return api.StatusNone, fmt.Errorf("unknown EVSE state: %d", state)
+	}
+}
+
+// Enabled implements the api.Charger interface
+func (c *OpenEVSE) Enabled() (bool, error) {
+	ctx, cancel := c.requestContextWithTimeout()
+	defer cancel()
+	overrideResp, err := c.api.GetManualOverrideWithResponse(ctx)
+
+	if err != nil {
+		return false, err
+	}
+
+	if overrideResp.JSON200 != nil && overrideResp.JSON200.State != nil {
+		switch *overrideResp.JSON200.State {
+		case "disabled":
+			return false, nil
+		case "enabled", "active":
+			return true, nil
+		}
+	}
+
+	statusResp, err := c.api.GetStatusWithResponse(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	var stateCode int
+	if statusResp.JSON200 != nil && statusResp.JSON200.State != nil {
+		stateCode = *statusResp.JSON200.State
+	} else {
+		stateCode = -1
+	}
+
+	var state bool
+	switch stateCode {
+	case 3, 4:
+		state = true
+	default:
+		configResp, err := c.api.GetConfigWithResponse(ctx)
+		if err != nil {
+			return false, err
+		}
+
+		switch *configResp.JSON200.ChargeMode {
+		case "fast":
+			state = true
+		}
+	}
+
+	err = c.SetManualOverride(state)
+
+	return state, err
+}
+
+// Enable implements the api.Charger interface
+func (c *OpenEVSE) Enable(enable bool) error {
+	if enable && c.phases == 0 {
```

Comment:
I broke it here. What has phase switching to do in this function? If it's required my last change is not valid :/

Reply:
The EVSE can do phase switching on its own. I want to disable it so only EVCC controls 1/3 phase change.

Reply:
In that case we should disable the auto-switching at startup and not continuously check?

Reply:
That should be enough. I was overly cautious.